pdns/lua-record.cc

```diff
@@ -309,61 +311,165 @@ static std::string getGeo(const std::string& ip, GeoIPInterface::GeoIPQueryAttri
   }
-  return ips[dns_random(ips.size())];
+  return items[dns_random(items.size())];
 }
 
-static ComboAddress hashed(const ComboAddress& who, const vector<ComboAddress>& ips)
+static ComboAddress pickRandomComboAddress(const vector<ComboAddress>& items)
 {
-  if (ips.empty()) {
-    throw std::invalid_argument("The IP list cannot be empty");
+  if (items.empty()) {
+    throw std::invalid_argument("The items list cannot be empty");
+  }
+  return items[dns_random(items.size())];
+}
+
+static string pickHashedString(const ComboAddress& who, const vector<string>& items)
+{
+  if (items.empty()) {
+    throw std::invalid_argument("The items list cannot be empty");
   }
   ComboAddress::addressOnlyHash aoh;
-  return ips[aoh(who) % ips.size()];
+  return items[aoh(who) % items.size()];
 }
 
+static ComboAddress pickHashedComboAddress(const ComboAddress& who, const vector<ComboAddress>& items)
+{
+  if (items.empty()) {
+    throw std::invalid_argument("The items list cannot be empty");
+  }
+  ComboAddress::addressOnlyHash aoh;
+  return items[aoh(who) % items.size()];
+}
 
-static ComboAddress pickwrandom(const vector<pair<int,ComboAddress> >& wips)
+static string pickWeightedRandomString(const vector< pair<int, string> >& items)
 {
-  if (wips.empty()) {
-    throw std::invalid_argument("The IP list cannot be empty");
+  if (items.empty()) {
+    throw std::invalid_argument("The items list cannot be empty");
   }
   int sum=0;
-  vector<pair<int, ComboAddress> > pick;
-  for(auto& i : wips) {
+  vector< pair<int, string> > pick;
+  pick.reserve(items.size());
+
+  for(auto& i : items) {
     sum += i.first;
     pick.emplace_back(sum, i.second);
   }
+  
+  if (sum == 0) {
+    /* we should not have any weight of zero, but better safe than sorry */
+    return std::string();
+  }
+  
   int r = dns_random(sum);
   auto p = upper_bound(pick.begin(), pick.end(), r, [](int rarg, const decltype(pick)::value_type& a) { return rarg < a.first; });
   return p->second;
 }
 
-static ComboAddress pickwhashed(const ComboAddress& bestwho, vector<pair<int,ComboAddress> >& wips)
+static ComboAddress pickWeightedRandomComboAddress(const vector< pair<int, ComboAddress> >& items)
 {
-  if (wips.empty()) {
+  if (items.empty()) {
+    throw std::invalid_argument("The items list cannot be empty");
+  }
+  int sum=0;
+  vector< pair<int, ComboAddress> > pick;
+  pick.reserve(items.size());
+
+  for(auto& i : items) {
+    sum += i.first;
+    pick.emplace_back(sum, ComboAddress(i.second));
+  }
+  
+  if (sum == 0) {
+    /* we should not have any weight of zero, but better safe than sorry */
     return ComboAddress();
   }
+  
+  int r = dns_random(sum);
+  auto p = upper_bound(pick.begin(), pick.end(), r, [](int rarg, const decltype(pick)::value_type& a) { return rarg < a.first; });
+  return p->second;
+}
+
+static string pickWeightedHashedString(const ComboAddress& bestwho, vector< pair<int, string> >& items)
```

Comment:
the duplication between the two variants of this function is too much - maybe we can template it? (also goes for a few others)

Reply:
Updated with the template-based approach as discussed